app/services/external_upload_manager.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+class ExternalUploadManager
+  DOWNLOAD_LIMIT = 100.megabytes
+
+  class ChecksumMismatchError < StandardError; end
+  class DownloadFailedError < StandardError; end
+  class CannotPromoteError < StandardError; end
+
+  attr_reader :external_upload_stub
+
+  def initialize(external_upload_stub)
+    @external_upload_stub = external_upload_stub
+  end
+
+  def can_promote?
+    external_upload_stub.status == ExternalUploadStub.statuses[:created]
+  end
+
+  def promote_to_upload!
+    raise CannotPromoteError if !can_promote?
+
+    external_upload_stub.update!(status: ExternalUploadStub.statuses[:uploaded])
```

Comment:
Should we only update the status only when the S3 actions have completed successfully?

Reply:
@tgxworld The `uploaded` status indicates it has been uploaded to S3, so we do this straight away to reflect that state, and to prevent 2 requests trying to process this upload at the same time (hence the can_promote?) check

Reply:
Ahh icic got it :+1: